packages/federation/src/managed-federation.ts

```diff
@@ -0,0 +1,439 @@
+   * Local fetch errors are thrown as exceptions.
+   */
+  error: { code: string; message: string };
+  /**
+   * The minimum delay in seconds to wait before trying to fetch this supergraph again.
+   */
+  minDelaySeconds: number;
+};
+
+type RouterConfigResult = {
+  routerConfig:
+    | {
+        __typename: 'RouterConfigResult';
+        messages: { level: 'ERROR' | 'WARN' | 'INFO'; body: string }[];
+        supergraphSDL: string;
+        minDelaySeconds: number;
+        id: string;
+      }
+    | { __typename: 'Unchanged'; minDelaySeconds: number; id: string }
+    | { __typename: 'FetchError'; code: string; message: string; minDelaySeconds: never };
+};
+
+/**
+ * The default managed federation up links. In case of failure, you should try to cycle through these up links.
+ *
+ * The first one is Apollo's managed federation up link on GCP, the second one is on AWS.
+ */
+export const DEFAULT_UPLINKS = [
+  'https://uplink.api.apollographql.com/',
+  'https://aws.uplink.api.apollographql.com/',
+];
+
+/**
+ * Fetches the supergraph SDL from a managed federation GraphOS up link.
+ * @param options
+ * @throws When the fetch fails or the response is not a valid.
+ * @returns An object with the supergraph SDL when possible. It also includes metadata to handle polling and retry logic.
+ *
+ *          If `lastSeenId` is provided and the supergraph has not changed, `supergraphSdl` is not present.
+ *
+ *          If The up link report a fetch error (which is not a local fetch error), it will be returned along with polling/retry metadata.
+ *          Any local fetch error will be thrown as an exception.
+ */
+export async function fetchSupergraphSdlFromManagedFederation(
+  options: FetchSupergraphSdlFromManagedFederationOpts = {},
+): Promise<RouterConfig | Unchanged | FetchError> {
+  const userDefinedUplinks =
+    process.env['APOLLO_SCHEMA_CONFIG_DELIVERY_ENDPOINT']?.split(',') ?? [];
+
+  const {
+    upLink = userDefinedUplinks[0] || DEFAULT_UPLINKS[0],
+    loggerByMessageLevel = DEFAULT_MESSAGE_LOGGER(console),
+    fetch = defaultFetch,
+    ...variables
+  } = options;
+
+  if (!variables.graphRef) {
+    variables.graphRef = process.env['APOLLO_GRAPH_REF'];
+  }
+
+  if (!variables.apiKey) {
+    variables.apiKey = process.env['APOLLO_KEY'];
+  }
+
+  const response = await fetch(upLink, {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json',
+    },
+    body: JSON.stringify({
+      query: /* GraphQL */ `
+        query ($apiKey: String!, $graphRef: String!, $lastSeenId: ID) {
+          routerConfig(ref: $graphRef, apiKey: $apiKey, ifAfterId: $lastSeenId) {
+            __typename
+            ... on FetchError {
+              code
+              message
+              minDelaySeconds
+            }
+            ... on Unchanged {
+              id
+              minDelaySeconds
+            }
+            ... on RouterConfigResult {
+              id
+              supergraphSdl: supergraphSDL
```

Comment:
```suggestion
              supergraphSDL
```
Let's not forget this one then if use `supergraphSDL` in the result object.

Reply:
Aaaah ok I didn't see you've added a label here, will revert my change in the result type and update types instead I think